src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,1840 @@
+/** -*- C++ -*- *****************************************************\
+
+Module: Constraint generation from string function calls
+        for the PASS algorithm (see the PASS paper at HVC'13)
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#include <solvers/refinement/string_constraint_generator.h>
+#include <util/unicode.h>
+#include <ansi-c/string_constant.h>
+#include <solvers/floatbv/float_bv.h>
+#include <java_bytecode/java_types.h>
+#include <util/pointer_predicates.h>
+
+constant_exprt string_constraint_generatort::constant_char(int i)
+{
+  switch(language)
+    {
+    case C :
+      return constant_exprt(integer2binary(i,STRING_SOLVER_CHAR_WIDTH),refined_string_typet::char_type());
+      break;
+    case JAVA : return constant_exprt(integer2binary(i,JAVA_STRING_SOLVER_CHAR_WIDTH),refined_string_typet::java_char_type());
+      break;
+    default: assert(false);
+    }
+}
+
+
+constant_exprt string_constraint_generatort::constant_unsigned(int i, size_t width)
+{ return constant_exprt(integer2binary(i,width),unsignedbv_typet(width)); }
+
+constant_exprt string_constraint_generatort::constant_signed(int i, size_t width)
+{ return constant_exprt(integer2binary(i,width),signedbv_typet(width)); }
+
+void string_constraint_generatort::check_char_type(const exprt & str)
+{
+  if(language == C)
+    assert(refined_string_typet::is_c_string_type(str.type()));
+  else
+    if(language == UNKNOWN)
+      {
+	if(refined_string_typet::is_c_string_type(str.type()))
+	  language = C;
+	else
+	  language = JAVA;
+      }
+
+}
+
+unsignedbv_typet string_constraint_generatort::get_char_type()
+{
+  if(language==C)
+    return refined_string_typet::char_type();
+  else if(language==JAVA) return refined_string_typet::java_char_type();
+  else assert(false);
+}
+
+size_t string_constraint_generatort::get_char_width()
+{
+  if(language==C)
+    return STRING_SOLVER_CHAR_WIDTH;
+  else if(language==JAVA) return JAVA_STRING_SOLVER_CHAR_WIDTH;
+  else assert(false);
+}
+
+symbol_exprt string_constraint_generatort::fresh_univ_index(const irep_idt &prefix)
+{
+  return string_exprt::fresh_symbol(prefix,refined_string_typet::index_type());
+}
+
+symbol_exprt string_constraint_generatort::fresh_exist_index(const irep_idt &prefix)
+{
+  symbol_exprt s = string_exprt::fresh_symbol(prefix,refined_string_typet::index_type());
+  index_symbols.push_back(s);
+  return s;
+}
+
+symbol_exprt string_constraint_generatort::fresh_boolean(const irep_idt &prefix)
+{
+  symbol_exprt b = string_exprt::fresh_symbol(prefix,bool_typet());
+  boolean_symbols.push_back(b);
+  return b;
+}
+
+
+string_exprt string_constraint_generatort::string_of_expr(const exprt & unrefined_string)
```

Comment:
Done in https://github.com/diffblue/cbmc/pull/341/commits/91ebe6198ca8f6d5d33c644f48eb914f0199f89e